Python/ast_opt.c

```diff
@@ -1005,6 +1012,29 @@ astfold_stmt(stmt_ty node_, PyArena *ctx_, _PyASTOptimizeState *state)
     case Assert_kind:
         CALL(astfold_expr, expr_ty, node_->v.Assert.test);
         CALL_OPT(astfold_expr, expr_ty, node_->v.Assert.msg);
+        /* Always emit a warning if the test is a non-zero length tuple */
+        if ((node_->v.Assert.test->kind == Tuple_kind &&
+            asdl_seq_LEN(node_->v.Assert.test->v.Tuple.elts) > 0) ||
+            (node_->v.Assert.test->kind == Constant_kind &&
+             PyTuple_Check(node_->v.Assert.test->v.Constant.value) &&
+             PyTuple_Size(node_->v.Assert.test->v.Constant.value) > 0))
+        {
+            PyObject *msg = PyUnicode_FromString("assertion is always true, "
+                                                 "perhaps remove parentheses?");
+            if (msg == NULL) {
+                return 0;
+            }
+            int ret = _PyErr_EmitSyntaxWarning(msg, state->filename,
+                                               node_->lineno, node_->col_offset + 1,
+                                               node_->end_lineno, node_->end_col_offset + 1);
+            Py_DECREF(msg);
+            if (ret < 0) {
+                return 0;
+            }
+        }
+        if (state->optimize) {
+            make_pass(node_);
```

Comment:
Technically speaking, this is not the removal of the `assert` statement; it is a replacement for the `pass` statement. Is it possible to completely remove the `assert node? Or will it break some existing code?

Reply:
Pass becomes a NOP, which gets removed later. This is how we get rid of code. However, I did forget to wipe out the location information (a NOP is not removed if it's the only instruction from its line in the source code).